src/components/teleporter.js

```diff
@@ -0,0 +1,377 @@
+    side: THREE.DoubleSide,
+    transparent: true
+  });
+
+  this.mesh = new THREE.Mesh(this.geometry, this.material);
+  this.mesh.drawMode = THREE.TriangleStripDrawMode;
+
+  this.mesh.frustumCulled = false;
+  this.mesh.vertices = this.vertices;
+
+  this.direction = new THREE.Vector3();
+  this.numPoints = numPoints;
+};
+
+RayCurve.prototype = {
+  setDirection: function(direction) {
+    const UP = new THREE.Vector3(0, 1, 0);
+    this.direction
+      .copy(direction)
+      .cross(UP)
+      .normalize()
+      .multiplyScalar(this.width / 2);
+  },
+
+  setWidth: function(width) {
+    this.width = width;
+  },
+
+  setPoint: (function() {
+    const posA = new THREE.Vector3();
+    const posB = new THREE.Vector3();
+
+    return function(i, point) {
+      posA.copy(point).add(this.direction);
+      posB.copy(point).sub(this.direction);
+
+      let idx = 2 * 3 * i;
+      this.vertices[idx++] = posA.x;
+      this.vertices[idx++] = posA.y;
+      this.vertices[idx++] = posA.z;
+
+      this.vertices[idx++] = posB.x;
+      this.vertices[idx++] = posB.y;
+      this.vertices[idx++] = posB.z;
+
+      this.geometry.attributes.position.needsUpdate = true;
+    };
+  })()
+};
+
+function parabolicCurve(p0, v0, t, out) {
+  out.x = p0.x + v0.x * t;
+  out.y = p0.y + v0.y * t - 4.9 * t * t;
+  out.z = p0.z + v0.z * t;
+  return out;
+}
+
+function isValidNormalsAngle(collisionNormal, referenceNormal, landingMaxAngle) {
+  const angleNormals = referenceNormal.angleTo(collisionNormal);
+  return THREE.Math.RAD2DEG * angleNormals <= landingMaxAngle;
+}
+
+const checkLineIntersection = (function() {
+  const direction = new THREE.Vector3();
+  return function checkLineIntersection(start, end, meshes, raycaster, referenceNormal, landingMaxAngle, hitPoint) {
+    direction.copy(end).sub(start);
+    const distance = direction.length();
+    raycaster.far = distance;
+    raycaster.set(start, direction.normalize());
+    const intersects = raycaster.intersectObjects(meshes, true);
+    if (intersects.length > 0 && isValidNormalsAngle(intersects[0].face.normal, referenceNormal, landingMaxAngle)) {
+      hitPoint.copy(intersects[0].point);
+      return true;
+    }
+    return false;
+  };
+})();
+
+function getMeshes(collisionEntities) {
+  return collisionEntities
+    .map(function(entity) {
+      return entity.getObject3D("mesh");
+    })
+    .filter(function(n) {
+      return n;
+    });
+}
+
+const MISS_OPACITY = 0.1;
+const HIT_OPACITY = 0.3;
+const MISS_COLOR = 0xff0000;
+const HIT_COLOR = 0x00ff00;
+const FORWARD = new THREE.Vector3(0, 0, -1);
+const LANDING_NORMAL = new THREE.Vector3(0, 1, 0);
+const MAX_LANDING_ANGLE = 45;
+const DRAW_TIME_MS = 400;
+const q = new THREE.Quaternion();
+const vecHelper = new THREE.Vector3();
+
+AFRAME.registerComponent("teleporter", {
+  schema: {
+    start: { type: "string" },
+    confirm: { type: "string" },
+    speed: { default: 12 },
+    collisionEntities: { default: "" },
```

Comment:
We only care about navmesh and we end up having a whole component dedicated to getting this to look at navmesh at the right time, I would vote we just do it more directly and less configurable.